### PR TITLE
@tko/observable) begin typescriptification

### PR DESCRIPTION
**Pull request by brianmhunt**

### CODE REVIEW

**Comment by brianmhunt** [comment]
@caseyWebb @Sebazzz  — here's where I'm working on the (likely quite naïve) typescriptification of Observables.  Any feedback and guidance is most welcome.

File: packages/observable/src/subscribable.ts [hunk 1]
```diff
@@ -3,38 +3,40 @@ import {
...
   [Symbol.observable] () { return this },
 
-  init (instance) {
+  init<T> (instance : Subscribable<T>) : void {
     instance._subscriptions = { change: [] }
     instance._versionNumber = 1
   },
 
-  subscribe (callback, callbackTarget, event) {
+  subscribe<T> (this: Subscribable<T>, callback: Subscriber, callbackTarget?: Function, event = defaultEvent) : Subscription {
```
Comment (caseyWebb) [inline packages/observable/src/subscribable.ts]: Avoid `Function` types. In fact, it should probably be added to tslint; I believe there is a `ban-types` option.

File: packages/observable/src/subscribable.ts [hunk 2]
```diff
@@ -3,38 +3,40 @@ import {
...
 
-export var defaultEvent = 'change'
+interface TC39Callback {
+}
+type Subscriber = (value: any) => void | TC39Callback
 
-var ko_subscribable_fn = {
+export const defaultEvent = 'change'
+
+const ko_subscribable_fn = {
```
Comment (caseyWebb) [inline packages/observable/src/subscribable.ts]: By defining all of the types in an `interface` or `type` (e.g. `ISubscribable<T>`) and then assigning `ko_subscribable_fn` to that (i.e. `ko_subscribable_fn: ISubscribable<T>`), `T` becomes accessible to all of the members below (almost all of the `any`s below are actually `T`). `this` can by typed as well if desired, and `ISubscribable<T>` becomes reusable.

**Review by caseyWebb — CHANGES_REQUESTED** [review]

File: packages/observable/src/subscribable.ts (hunk 2 above)
Comment (caseyWebb) [reply packages/observable/src/subscribable.ts]: I just noticed it's already defined and used in Observable. Just move it here and consolidate and it's good.

**Review by caseyWebb — CHANGES_REQUESTED** [review]

File: packages/observable/src/subscribable.ts [hunk 3]
```diff
@@ -3,38 +3,40 @@ import {
...
 
-export function subscribable () {
+export function subscribable<T> (this: Subscribable<T>) : void {
   Object.setPrototypeOf(this, ko_subscribable_fn)
   ko_subscribable_fn.init(this)
 }
 
-export var defaultEvent = 'change'
+interface TC39Callback {
+}
```
Comment (Sebazzz) [inline packages/observable/src/subscribable.ts:23]: Why the empty interface?

File: packages/observable/src/observable.ts [hunk 4]
```diff
@@ -8,12 +8,14 @@ import {
...
+import { subscribable, defaultEvent, LATEST_VALUE } from './subscribable'
 import { valuesArePrimitiveAndEqual } from './extenders.js'
 
-export function observable (initialValue) {
-  function Observable () {
-    if (arguments.length > 0) {
+interface Observable = (...args: any[]) => void
+
+export function observable (initialValue: any) : Observable {
+  function Observable (...args: any[]) : Observable {
```
Comment (Sebazzz) [inline packages/observable/src/observable.ts]: Good candidate to make generic. `Observable<T>` with `initialValue` of type `T`.

File: packages/observable/src/subscribable.ts (hunk 3 above)
Comment (brianmhunt) [reply packages/observable/src/subscribable.ts:23]: Just a TODO

File: packages/observable/src/subscribable.ts (hunk 2 above)
Comment (brianmhunt) [reply packages/observable/src/subscribable.ts]: @caseyWebb When I try
```ts
const ko_subscribable_fn: Subscribable<T> = {
```
I get a `Cannot find name T`.  Staring at it now. :)

**Comment by caseyWebb** [comment]
whoops, I didn't mean to do that, but no harm. if you push new commits I believe I can re-open.

File: packages/observable/src/subscribable.ts (hunk 2 above)
Comment (caseyWebb) [reply packages/observable/src/subscribable.ts]: @brianmhunt see this commit https://github.com/caseyWebb/tko/commit/2096642c83b8310bbd4029b4bc5434fe67e51455

**Comment by brianmhunt** [comment]
> whoops, I didn't mean to do that, but no harm. if you push new commits I believe I can re-open.

Worst case I'll add a new PR

**Pull request merged**